File: gcpc2026/garbledgarden/submissions/accepted/yidi.py
import sys
sys.setrecursionlimit(10000)

def add(a, x):
    for i in range(len(a)):
        a[i] += x

cycle = []
def euler(u):
    while adj[u]:
        v, x = adj[u].pop()
        euler(v)
        cycle.append(x)

n = int(input())
a = [*map(int, input().split())]
add(a, -1)

a2 = sorted(a)
adj = [[] for i in range(n)]
for it in range(2):
    rev = [0] * n
    for i in range(n):
        if a[i] != a2[i]:
            adj[a[i]].append((a2[i], i))
            rev[a[i]] = i
    
    comps = []
    cycle = []
    for i in range(n):
        if adj[i]:
            comps.append(rev[i])
            euler(i)
    cycle.reverse()

    if not comps:
        print(0)
        exit(0)

    if len(comps) == 1:
        if it == 0:
            print(1)
        print(len(cycle))
        add(cycle, 1)
        print(*cycle)
        exit(0)

    a3 = a[:]
    for i in range(len(comps)):
        a[comps[i]] = a3[comps[(i+1)%len(comps)]]
    
    print(2)
    print(len(comps))
    add(comps,1)
    print(*comps)

File: gcpc2026/garbledgarden/submissions/accepted/Yidi.java
import java.util.*;

public class Yidi {
    public static ArrayList<Integer> g[];
    public static ArrayList<Integer> cycle;

	public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        int[] a = new int[n];
        for (int i=0; i<n; i++) a[i] = sc.nextInt() - 1;

        int[] a2 = a.clone();
        Arrays.sort(a2);
        g = new ArrayList[n];
        for (int i=0; i<n; i++) g[i] = new ArrayList<Integer>();
        for (int it=0; it<2; it++) {
            int[] rev = new int[n];
            for (int i=0; i<n; i++) if (a[i] != a2[i]) {
                g[a[i]].add(a2[i]);
                g[a[i]].add(i);
                rev[a[i]] = i;
            }

            ArrayList<Integer> comps = new ArrayList<Integer>();
            cycle = new ArrayList<Integer>();
            for (int i=0; i<n; i++) if (!g[i].isEmpty()) {
                comps.add(rev[i]);
                euler(i);
            }
            Collections.reverse(cycle);

            if (comps.isEmpty()) {
                System.out.println(0);
                System.exit(0);
            }

            if (comps.size() == 1) {
                if (it == 0) {
                    System.out.println(1);
                }
                System.out.println(cycle.size());
                for (int e : cycle) System.out.print((e+1) + " ");
                System.out.println();
                System.exit(0);
            }

            int[] a3 = a.clone();
            for (int i=0; i<comps.size(); i++) {
                a[comps.get(i)] = a3[comps.get((i+1) % comps.size())];
            }

            System.out.println(2);
            System.out.println(comps.size());
            for (int e : comps) System.out.print((e+1) + " ");
            System.out.println();
        }
	}

    public static void euler(int u) {
        while (!g[u].isEmpty()) {
            int v = g[u].remove(0); // peak efficiency
            int x = g[u].remove(0);
            euler(v);
            cycle.add(x);
        }
    }
}
